### PR TITLE
Bricks: `random.h`

### PR DESCRIPTION
**Pull request by mzhurovich**
@dkorolev What do you think?

### CODE REVIEW

File: Bricks/util/random.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+inline void SetSeed(const size_t seed) {
+  ThreadLocalSingleton<impl::SeedImpl>().seed = seed;
+  ThreadLocalSingleton<impl::SeedImpl>().is_set = true;
+}
+
+inline std::mt19937_64& mt19937_64() { return ThreadLocalSingleton<impl::mt19937_64_wrapper>().instance; }
+
+template <typename T>
+inline typename std::enable_if<std::is_integral<T>::value, T>::type RandomInt(const T a, const T b) {
```
Comment (dkorolev) [inline Bricks/util/random.h]: Why `std::enable_if` if can just `static_assert()` in the body?

File: Bricks/util/test.cc [hunk 2]
```diff
@@ -230,3 +233,10 @@ TEST(Util, SHA256) {
   EXPECT_EQ("a591a6d40bf420404a011733cfb7b190d62c65bf0bcda32b57b277d9ad9f146e",
             static_cast<std::string>(bricks::SHA256("Hello World")));
 }
+
+TEST(Util, RandomWithFixedSeed) {
+  EXPECT_EQ(114, bricks::random::RandomInt<int>(-100, 200));
```
Comment (dkorolev) [inline Bricks/util/test.cc]: Oops on Ubuntu.

```
[ RUN      ] Util.RandomWithFixedSeed
test.cc:238: Failure
Value of: bricks::random::RandomInt<int>(-100, 200)
  Actual: 127
Expected: 114
test.cc:239: Failure
Value of: bricks::random::RandomInt<uint64_t>(1e10, 1e18)
  Actual: 654891597639083045
Expected: 258833541435025064u
Which is: 258833541435025064
```

File: Bricks/util/random.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  std::uniform_int_distribution<T> distribution(a, b);
+  return distribution(mt19937_64());
+}
+
+template <typename T>
+inline typename std::enable_if<std::is_floating_point<T>::value, T>::type RandomFloat(const T a, const T b) {
```
Comment (dkorolev) [inline Bricks/util/random.h]: I'd do with `Double` or `Real` in the name.

File: Bricks/util/random.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+}  // namespace impl
+
+inline void SetSeed(const size_t seed) {
+  ThreadLocalSingleton<impl::SeedImpl>().seed = seed;
+  ThreadLocalSingleton<impl::SeedImpl>().is_set = true;
+}
+
+inline std::mt19937_64& mt19937_64() { return ThreadLocalSingleton<impl::mt19937_64_wrapper>().instance; }
```
Comment (dkorolev) [inline Bricks/util/random.h]: Thought: I was puzzled that perhaps there's a better name for this method, that involves the word `singleton` or `thread_local`. Couldn't find any.

Perhaps the namespace could be called differently to reflect it? Ex. `namespace thread_local_random`?

File: Bricks/util/test.cc (hunk 2 above)
Comment (mzhurovich) [inline Bricks/util/test.cc]: Looks like we'll have to remove the test :(
http://stackoverflow.com/questions/26663140/c-normal-distribution-gives-different-results-on-different-platforms

**Comment by dkorolev** [comment]
I think our OCD-s are aligned :+1: 

PTAL at minor comments in the code; LGTM code-wise otherwise.

And we need to understand why are the numbers out of sync on Mac and Linux. :-( not before I have breakfast though.

File: Bricks/util/random.h (hunk 4 above)
Comment (mzhurovich) [inline Bricks/util/random.h]: Easy!
How about `mt19937_64_tls`?

Comment (mzhurovich) [inline Bricks/util/random.h]: I think this abbreviation is more or less common.

Comment (dkorolev) [inline Bricks/util/random.h]: like!

On Sunday, June 14, 2015, Maxim Zhurovich <notifications@github.com> wrote:

> In Bricks/util/random.h
> <https://github.com/c5t/Current/pull/212#discussion_r32382963>:
>
> > +#else
> > +  mt19937_64_wrapper()
> > +      : instance(ThreadLocalSingleton<SeedImpl>().is_set ? ThreadLocalSingleton<SeedImpl>().seed
> > +                                                         : static_cast<size_t>(bricks::time::Now())) {}
> > +#endif
> > +  std::mt19937_64 instance;
> > +};
> > +
> > +}  // namespace impl
> > +
> > +inline void SetSeed(const size_t seed) {
> > +  ThreadLocalSingleton<impl::SeedImpl>().seed = seed;
> > +  ThreadLocalSingleton<impl::SeedImpl>().is_set = true;
> > +}
> > +
> > +inline std::mt19937_64& mt19937_64() { return ThreadLocalSingleton<impl::mt19937_64_wrapper>().instance; }
>
> Easy!
> How about mt19937_64_tls?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/c5t/Current/pull/212/files#r32382963>.
>


-- 
And, as always, have a nice day!
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783

**Comment by mzhurovich** [comment]
@dkorolev made changes and temporarily disabled the test.

**Comment by dkorolev** [comment]
LGTM && Merging. Go OCD! :+1:

**Comment by mzhurovich** [comment]
@dkorolev Yay! :)

**Pull request merged**